unixfs/src/file/visit.rs

```diff
@@ -0,0 +1,278 @@
+use cid::Cid;
+use std::borrow::Cow;
+use std::convert::TryFrom;
+use std::ops::Range;
+
+use crate::file::reader::{FileContent, FileReader, Traversal};
+use crate::file::{FileMetadata, FileReadFailed, UnwrapBorrowedExt};
+use crate::pb::merkledag::PBLink;
+
+/// IdleFileVisit represents a prepared file visit over a tree. The user has to know the CID and be
+/// able to get the block for the visit.
+#[derive(Default, Debug)]
+pub struct IdleFileVisit {
+    range: Option<Range<u64>>,
+}
+
+impl IdleFileVisit {
+    /// Target range represents the target byte range of the file we are interested in visiting.
+    pub fn with_target_range(self, range: Range<u64>) -> Self {
+        Self { range: Some(range) }
+    }
+
+    /// Begins the visitation by processing the first block to be visited.
+    ///
+    /// Returns on success a tuple of file bytes, any metadata associated, and optionally a
+    /// `FileVisit` to continue the walk.
+    pub fn start(
+        self,
+        block: &[u8],
+    ) -> Result<(&[u8], FileMetadata, Option<FileVisit>), FileReadFailed> {
+        let fr = FileReader::from_block(block)?;
+
+        let metadata = fr.as_ref().to_owned();
+
+        let (content, traversal) = fr.content();
+
+        match content {
+            FileContent::Just(content) => {
+                let block = 0..content.len() as u64;
+                let content = maybe_target_slice(content, &block, self.range.as_ref());
+                Ok((content, metadata, None))
+            }
+            FileContent::Spread(iter) => {
+                // we need to select suitable here
+                let mut pending = iter
+                    .enumerate()
+                    .filter_map(|(i, (link, range))| {
+                        if !block_is_in_target_range(&range, self.range.as_ref()) {
+                            return None;
+                        }
+
+                        Some(to_pending(i, link, range))
+                    })
+                    .collect::<Result<Vec<(Cid, Range<u64>)>, _>>()?;
```

Comment:
this might be hard to, so I'd think of it as a post-PR nice-to-have, but looking at how `pending` is used I _think_ it is possible to _not_ collect this iterator, even when later stuff is `push`ed to it (there's `chain` that could, with some luck, be applied there)

Reply:
This iterator is built from by zipping `Vec<PBLink<'_>>` and `Vec<u64>`. Not collecting here isn't really an option, because we would not have anything `'static` and without it a buffer could not be reused. The Cid transformation could be done earlier but I think storing the chained iterators would require boxing on top of boxing. With iterator chaining this could probably be `Vec<(Cid, u64)>` (transformed PBLink and it's blocksize), and an `u64` start offset.

The `Range<u64>` next to a `Cid` does probably waste bytes though. It might be reasonable just to store the tuple in two Vecs in SoA fashion.

The strategy for memory saving I've been thinking about would be to set an upper bound, and somehow store the "pending" links as relative to the `Cid` are recorded in with two offsets where the last link and last blocksize ended in, something like `(Cid, u64, u64)`. When we would come back to such link after traversing nodes below it, the `Block` could be loaded back up from the disk and the only the next link and it's blocksize parsed.

I would had originally pursued that avenue with `minipb` but then I learned that the default number of links is really 174 per block. Had it been 6k links it might be reasonable to batch load some configurable amount of them at once.

Reply:
> I learned that the default number of links is really 174 per block

This is a legacy default. That is - it is almost guaranteed that it will not be part of the next iteration of unixfs. I would **very strongly** urge you to not design around this number.

Reply:
> > I learned that the default number of links is really 174 per block
> 
> This is a legacy default.

Agreed.

> I would very strongly urge you to not design around this number.

Thanks! Luckily this consideration at the moment this mostly affects whether or not to spend few weeks more on writing the block re-reading scheme I described above. I had not thought about it but maybe this visitation could be usable for cbor as well, even the same strategy I visioned for memory savings above would/could apply.

Reply:
I did look into separating the storage of `Vec<(Cid, Range<u64>)` but that did not show up in major changes in `massif` or `dhat` output of running the `cat` example with output to `/dev/null`. There was a surprising allocation whenever opening a file in `std` but while interesting, it doesn't really matter either.